Demeter smart contract upgrades

arcantheon opened the pull request.

YashP16 committed:
feat(contracts/Farm): Add first draft for generic Fungible token farm.

YashP16 committed:
feat(contracts/UniV2FarmDeployer): Add generic UniV2 farm deployer

* Deployer accepts generic pool parameters to create a farm on top of the pool extend from uniV2.
* We can have a deployer for any protocol extending uniV2 factory interface.

YashP16 committed:
feat(scripts): Add following deployment configurations:

* SushiSwap
* TraderJoe
* Camelot

YashP16 committed:
fix(scripts): Error while creating farm.

YashP16 committed:
Squashed commit of the following:

commit 6f9c3277b8e7da6c4e2e813e1d34da2dddccef52
Merge: 1e9b621 b0fdba1
Author: Yash Pitroda <yash.pitroda786@gmail.com>
Date:   Fri May 12 19:07:34 2023 +0530

    Merge pull request #8 from Sperax/wip/camelot-farm

    Add Demeter <> Camelot farm.

commit b0fdba1a5bccdebd9c1c1ceca7d0748fd0912bd2
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Fri May 12 18:55:30 2023 +0530

    docs: Add deployment artifacts for Camelot Farm.

commit cfaadc4ab67a39992a56327d395ce39a64b36dcc
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Fri May 12 17:26:56 2023 +0530

    fix(contracts/FarmDeployer): Update discountedFee.

commit 4b7a804032e7512757aa9ad584e46247a3fa8fd4
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Fri May 12 16:27:31 2023 +0530

    fix(scripts): Add farm_factory upgrade step in setup_script.

commit 5ff48243d647317a527e5da8ef5f9893df394ecf
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Fri May 12 14:24:03 2023 +0530

    refactor: Resolve review comments.

commit b5fd87f7243b5fd9d2cfa5f2b4850f7d2e881df1
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu May 11 20:49:16 2023 +0530

    refactor(contracts/Demeter_Camelot_Farm): Resolve todo comment.

commit aecc9a2886bdafc8098b3f4f3d754dd11bf74c29
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu May 11 20:43:32 2023 +0530

    refactor(contracts): Upgrade contract to v0.8.16

    * Resolve review comments.
    * Upgraded the version in brownie config.

commit 10481b52f47cb8728cbbcdaa86ea719918dd64be
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu May 11 15:04:18 2023 +0530

    refactor(contracts/FarmFactory): Reorder event args.

commit 6e90e8415b78b118940182cf6102ed8a8cba38e8
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu May 11 15:03:33 2023 +0530

    refactor(contracts/FarmFactory): Reorder event args.

commit 1b6f13fcd91be3d9eab71cab70e68c043bdb179e
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu May 11 14:59:59 2023 +0530

    revert(contracts/FarmFactory): Add the initializer back in FarmFactory.

commit bc9800fd58a3d9f00996c992e7fccc7b25e0326b
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu May 11 14:56:16 2023 +0530

    refactor(scripts): Add deployment configurations.

commit 48704826e781a577a1dbf4a009153406c8d18c9b
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu May 11 14:17:34 2023 +0530

    refactor: Save coverage reports in /reports

commit d3cb20a9c296137dddafe577a2cc1783c94528ba
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu May 11 14:15:09 2023 +0530

    Squashed commit of the following:

    commit 6a3f09b9a6920fd71005f2872f49fe6002c18867
    Author: YashP16 <yash.pitroda786@gmail.com>
    Date:   Thu May 11 14:13:31 2023 +0530

        test: update contract name

    commit 2ed07a24d0ad7464134207332217b225af277acb
    Merge: 92c93c8 b82f86f
    Author: YashP16 <yash.pitroda786@gmail.com>
    Date:   Thu May 11 13:45:55 2023 +0530

        Merge branch 'wip/camelot-farm' into test/camelot-farm

    commit 92c93c868e9c3cd711827ee7ef946bbbd016d7d8
    Author: Dayaa <m.d.e.boutalbi@me.com>
    Date:   Mon May 8 01:44:53 2023 +0100

         contract: Demeter_CamelotFarm - 97.5%
            BaseFarm._addRewardData - 100.0%
            BaseFarm._claimRewards - 100.0%
            BaseFarm._farmNotClosed - 100.0%
            BaseFarm._farmNotPaused - 100.0%
            BaseFarm._getAccRewards - 100.0%
            BaseFarm._initiateCooldown - 100.0%
            BaseFarm._isNonZeroAddr - 100.0%
            BaseFarm._isTokenManager - 100.0%
            BaseFarm._isValidCooldownPeriod - 100.0%
            BaseFarm._isValidDeposit - 100.0%
            BaseFarm._recoverRewardFunds - 100.0%
            BaseFarm._setRewardRate - 100.0%
            BaseFarm._setupFarm - 100.0%
            BaseFarm._subscribeRewardFund - 100.0%
            BaseFarm._unsubscribeRewardFund - 100.0%
            BaseFarm._updateFarmRewardData - 100.0%
            BaseFarm._withdraw - 100.0%
            BaseFarm.farmPauseSwitch - 100.0%
            BaseFarm.getRewardBalance - 100.0%
            BaseFarm.getRewardFundInfo - 100.0%
            BaseFarm.getSubscriptionInfo - 100.0%
            BaseFarm.recoverERC20 - 100.0%
            BaseFarm.updateFarmStartTime - 100.0%
            Demeter_CamelotFarm.onERC721Received - 100.0%
            Demeter_CamelotFarm.onNFTHarvest - 100.0%
            BaseFarm._deposit - 91.7%
            BaseFarm.computeRewards - 83.3%

    commit d50ac5c29efd4624375052d9c42026bd6c5558f1
    Merge: e9526c1 e31f790
    Author: Dayaa <m.d.e.boutalbi@me.com>
    Date:   Sun May 7 23:43:46 2023 +0100

        Merge branch 'test/camelot-farm' of github.com:Sperax/Demeter-Protocol into test/camelot-farm

    commit e9526c134d4d5b2fc23d5070a66a5a4974aa0926
    Author: Dayaa <m.d.e.boutalbi@me.com>
    Date:   Sun May 7 23:43:16 2023 +0100

          contract: Demeter_UniV3Farm_v2 - 97.1%
            BaseFarm._addRewardData - 100.0%
            BaseFarm._claimRewards - 100.0%
            BaseFarm._deposit - 100.0%
            BaseFarm._farmNotClosed - 100.0%
            BaseFarm._farmNotPaused - 100.0%
            BaseFarm._getAccRewards - 100.0%
            BaseFarm._initiateCooldown - 100.0%
            BaseFarm._isNonZeroAddr - 100.0%
            BaseFarm._isTokenManager - 100.0%
            BaseFarm._isValidCooldownPeriod - 100.0%
            BaseFarm._isValidDeposit - 100.0%
            BaseFarm._recoverRewardFunds - 100.0%
            BaseFarm._setRewardRate - 100.0%
            BaseFarm._setupFarm - 100.0%
            BaseFarm._subscribeRewardFund - 100.0%
            BaseFarm._unsubscribeRewardFund - 100.0%
            BaseFarm._updateFarmRewardData - 100.0%
            BaseFarm._withdraw - 100.0%
            BaseFarm.farmPauseSwitch - 100.0%
            BaseFarm.getRewardFundInfo - 100.0%
            BaseFarm.getSubscriptionInfo - 100.0%
            BaseFarm.recoverERC20 - 100.0%
            BaseFarm.updateFarmStartTime - 100.0%
            Demeter_UniV3Farm_v2._validateTickRange - 100.0%
            Demeter_UniV3Farm_v2.onERC721Received - 100.0%
            BaseFarm.getRewardBalance - 93.8%
            Demeter_UniV3Farm_v2._getLiquidity - 91.7%
            Demeter_UniV3Farm_v2.claimUniswapFee - 87.5%
            BaseFarm.computeRewards - 83.3%

    commit e31f7906b19f3f26aea92aa1a60ddbdcda03e52f
    Merge: 6bdf6e8 d24418e
    Author: YashP16 <yash.pitroda786@gmail.com>
    Date:   Wed May 3 12:31:53 2023 +0530

        Merge branch 'wip/camelot-farm' into test/camelot-farm

    commit 6bdf6e822c13ee5f81cee41860bceae5faae9508
    Merge: d076834 3c01197
    Author: YashP16 <yash.pitroda786@gmail.com>
    Date:   Tue May 2 19:59:39 2023 +0530

        Merge branch 'wip/camelot-farm' into test/camelot-farm

    commit d076834de3bbd5b5c7ad102f2fa9bfb630707672
    Author: YashP16 <yash.pitroda786@gmail.com>
    Date:   Tue May 2 19:39:27 2023 +0530

        fix(tests): Restore uniV3 farm tests.

    commit 50585d25d74d9c7449e42f48ff6f7452e74dd1ec
    Merge: e30c6fb a01a06b
    Author: YashP16 <yash.pitroda786@gmail.com>
    Date:   Tue May 2 12:13:47 2023 +0530

        Merge branch 'wip/camelot-farm' into test/camelot-farm

    commit e30c6fb343c2372310bde1e9ffbd11eba9f80b96
    Author: Dayaa <m.d.e.boutalbi@me.com>
    Date:   Thu Apr 13 03:42:53 2023 +0100

          contract: Demeter_CamelotFarm - 98.1%
            Demeter_CamelotFarm._addRewardData - 100.0%
            Demeter_CamelotFarm._claimRewards - 100.0%
            Demeter_CamelotFarm._farmNotClosed - 100.0%
            Demeter_CamelotFarm._farmNotPaused - 100.0%
            Demeter_CamelotFarm._getAccRewards - 100.0%
            Demeter_CamelotFarm._isNonZeroAddr - 100.0%
            Demeter_CamelotFarm._isTokenManager - 100.0%
            Demeter_CamelotFarm._isValidCooldownPeriod - 100.0%
            Demeter_CamelotFarm._isValidDeposit - 100.0%
            Demeter_CamelotFarm._recoverRewardFunds - 100.0%
            Demeter_CamelotFarm._setRewardRate - 100.0%
            Demeter_CamelotFarm._setupFarm - 100.0%
            Demeter_CamelotFarm._subscribeRewardFund - 100.0%
            Demeter_CamelotFarm._unsubscribeRewardFund - 100.0%
            Demeter_CamelotFarm._updateFarmRewardData - 100.0%
            Demeter_CamelotFarm.farmPauseSwitch - 100.0%
            Demeter_CamelotFarm.getRewardBalance - 100.0%
            Demeter_CamelotFarm.getRewardFundInfo - 100.0%
            Demeter_CamelotFarm.getSubscriptionInfo - 100.0%
            Demeter_CamelotFarm.initialize - 100.0%
            Demeter_CamelotFarm.initiateCooldown - 100.0%
            Demeter_CamelotFarm.onNFTHarvest - 100.0%
            Demeter_CamelotFarm.recoverERC20 - 100.0%
            Demeter_CamelotFarm.updateFarmStartTime - 100.0%
            Demeter_CamelotFarm.withdraw - 100.0%
            Demeter_CamelotFarm.onERC721Received - 95.8%
            Demeter_CamelotFarm.computeRewards - 91.7%

    commit 59d707768d0df791dcc64f570b4ec37b290723c4
    Author: Dayaa <m.d.e.boutalbi@me.com>
    Date:   Wed Apr 12 11:18:06 2023 +0100

          contract: Demeter_CamelotFarm - 97.6%
            Demeter_CamelotFarm._addRewardData - 100.0%
            Demeter_CamelotFarm._claimRewards - 100.0%
            Demeter_CamelotFarm._farmNotClosed - 100.0%
            Demeter_CamelotFarm._farmNotPaused - 100.0%
            Demeter_CamelotFarm._getAccRewards - 100.0%
            Demeter_CamelotFarm._isNonZeroAddr - 100.0%
            Demeter_CamelotFarm._isTokenManager - 100.0%
            Demeter_CamelotFarm._isValidCooldownPeriod - 100.0%
            Demeter_CamelotFarm._isValidDeposit - 100.0%
            Demeter_CamelotFarm._recoverRewardFunds - 100.0%
            Demeter_CamelotFarm._setRewardRate - 100.0%
            Demeter_CamelotFarm._setupFarm - 100.0%
            Demeter_CamelotFarm._subscribeRewardFund - 100.0%
            Demeter_CamelotFarm._unsubscribeRewardFund - 100.0%
            Demeter_CamelotFarm._updateFarmRewardData - 100.0%
            Demeter_CamelotFarm.farmPauseSwitch - 100.0%
            Demeter_CamelotFarm.getRewardBalance - 100.0%
            Demeter_CamelotFarm.getRewardFundInfo - 100.0%
            Demeter_CamelotFarm.getSubscriptionInfo - 100.0%
            Demeter_CamelotFarm.initialize - 100.0%
            Demeter_CamelotFarm.initiateCooldown - 100.0%
            Demeter_CamelotFarm.onNFTHarvest - 100.0%
            Demeter_CamelotFarm.recoverERC20 - 100.0%
            Demeter_CamelotFarm.updateFarmStartTime - 100.0%
            Demeter_CamelotFarm.withdraw - 100.0%
            Demeter_CamelotFarm.onERC721Received - 95.8%
            Demeter_CamelotFarm.computeRewards - 83.3%

    commit 436f24767411c3189afae77fb9b3e2871624588a
    Author: Dayaa <m.d.e.boutalbi@me.com>
    Date:   Wed Apr 12 01:55:39 2023 +0100

        Pool reward tests checked

    commit 100e82448a3ea167a5efeec754be134bb7898d84
    Merge: 79122f3 b4dcb21
    Author: Dayaa <m.d.e.boutalbi@me.com>
    Date:   Tue Apr 11 13:04:25 2023 +0100

        Merge branch 'wip/camelot-farm' into test/camelot-farm

    commit 79122f399a2520732ac3708810b740c9ee4af844
    Merge: e9ffc19 eb0c202
    Author: Dayaa <m.d.e.boutalbi@me.com>
    Date:   Mon Apr 10 10:13:05 2023 +0100

        Merge branch 'wip/camelot-farm' into test/camelot-farm

    commit e9ffc19b077bc359099c3ad8cf283f7120662be8
    Author: Dayaa <m.d.e.boutalbi@me.com>
    Date:   Mon Apr 10 10:11:37 2023 +0100

        Add Testing Claim Pool Rewards Function

    commit cf539015f20b99f0fc37de5e207e61695262768f
    Merge: 2eba21e 6de0223
    Author: Dayaa <m.d.e.boutalbi@me.com>
    Date:   Thu Apr 6 18:35:49 2023 +0100

        Merge branch 'wip/camelot-farm' into test/camelot-farm

    commit 2eba21e1ccaad9579a52570596b899dfb9670298
    Author: Dayaa <m.d.e.boutalbi@me.com>
    Date:   Thu Apr 6 18:34:12 2023 +0100

          contract: Demeter_CamelotFarm - 94.2%
            Demeter_CamelotFarm._addRewardData - 100.0%
            Demeter_CamelotFarm._claimRewards - 100.0%
            Demeter_CamelotFarm._farmNotClosed - 100.0%
            Demeter_CamelotFarm._farmNotPaused - 100.0%
            Demeter_CamelotFarm._isNonZeroAddr - 100.0%
            Demeter_CamelotFarm._isTokenManager - 100.0%
            Demeter_CamelotFarm._isValidCooldownPeriod - 100.0%
            Demeter_CamelotFarm._isValidDeposit - 100.0%
            Demeter_CamelotFarm._recoverRewardFunds - 100.0%
            Demeter_CamelotFarm._setRewardRate - 100.0%
            Demeter_CamelotFarm._setupFarm - 100.0%
            Demeter_CamelotFarm._updateFarmRewardData - 100.0%
            Demeter_CamelotFarm.farmPauseSwitch - 100.0%
            Demeter_CamelotFarm.getRewardBalance - 100.0%
            Demeter_CamelotFarm.getRewardFundInfo - 100.0%
            Demeter_CamelotFarm.getSubscriptionInfo - 100.0%
            Demeter_CamelotFarm.initialize - 100.0%
            Demeter_CamelotFarm.initiateCooldown - 100.0%
            Demeter_CamelotFarm.recoverERC20 - 100.0%
            Demeter_CamelotFarm.updateFarmStartTime - 100.0%
            Demeter_CamelotFarm.withdraw - 100.0%
            Demeter_CamelotFarm._unsubscribeRewardFund - 87.5%
            Demeter_CamelotFarm.onERC721Received - 87.5%
            Demeter_CamelotFarm.computeRewards - 83.3%
            Demeter_CamelotFarm._subscribeRewardFund - 75.0%
            Demeter_CamelotFarm._getAccRewards - 70.8%

    commit cc4ec1b0a60970598e523eaea2d364b5e0a7581b
    Author: Dayaa <m.d.e.boutalbi@me.com>
    Date:   Wed Apr 5 04:44:13 2023 +0100

         error onERC721Received: incorrect nft

commit b82f86fdea8333c0cf11c7e7541412e09dfcdf24
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu May 11 13:36:37 2023 +0530

    refactor(contracts/Demeter_UniV3Farm): Rename farm and deployer contract for consistency

commit afe98ee6da184979916a7b903cec6baa82bb69bf
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu May 11 13:35:08 2023 +0530

    refactor(contracts/BaseFarm): Add comments.

commit 11a90469bf4c3df6b0f983fd50467d06b95e9e98
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu May 11 12:55:04 2023 +0530

    refactor(contracts/FarmFactory): Move the privilegedUser logic to farmFactory contract.

commit b3f4bed1867183a6c558b70a068aee5a27b08d20
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu May 11 12:54:20 2023 +0530

    chore(pre-commit-hook): Add branch name restriction.

commit 949da79f4963be51da9d4c4e85e961d04bef8153
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Wed May 3 12:49:49 2023 +0530

    refactor(scripts): Update uniV3 farm deployer in the script.

commit d24418efa512e8ef805feb9f37cfb7d6e170ce95
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Tue May 2 23:09:36 2023 +0530

    refactor(contracts/FarmDeployer): Update event order for createFarm()

commit 4f491da3919a3386abab8f635b449ef562270f94
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Tue May 2 20:04:27 2023 +0530

    refactor(contracts/FarmFactory): Update FarmRegistered event to map FarmDeployer and FarmContract.

commit 3c01197f327278b428cf89b72df77db2dcc947ea
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Tue May 2 19:50:24 2023 +0530

    refactor(contracts): Create a BaseFarm contract abstracting out the common demeter farm logic.

commit a01a06b999002641557b9cb43c1f76fd7094ef69
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Fri Apr 28 16:49:53 2023 +0530

    refactor(contracts): Add farmId in Demeter_CamelotFarm.

commit b4dcb21ba3689ecb640e55e8a7027f81eff4bc94
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Tue Apr 11 17:30:30 2023 +0530

    refactor(contracts/Demeter_CamelotFarm): Capture reward information in the onNFTHarvest().

commit 9411d317f8eab59142400def4fa5aae8999c9ccb
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Tue Apr 11 17:12:19 2023 +0530

    refactor(contracts/BaseFarmDeployer): Remove redundant code blocks from deployer implementation

commit eb0c202fad2e33df20706e8ecdd8116db6af48ef
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Mon Apr 10 13:26:52 2023 +0530

    fix(contracts/Demeter_CamelotFarm): Make farm contract compatible to harvest camelot pool rewards.

commit 6de0223c3da40ebfe09ce6a9999a47d0d6fd85fe
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Mon Apr 3 15:13:35 2023 +0530

    refactor(contracts/Demeter_CamelotFarm): Resolve review comments.

commit 6bdf33202637efa9a0c5809a372606276809b46f
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu Mar 30 21:28:14 2023 +0530

    feat(scripts): Add setup and manual testing script for camelot farms.

commit cef0f462267948b5ac5f8d4355a045b70aac8882
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu Mar 30 21:27:23 2023 +0530

    refactor(scripts): Add deployment constants for camelot deployer and farm

commit 461b4d3fd71f516cd060f4c94f709bf47bc785fe
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Thu Mar 30 21:23:36 2023 +0530

    refactor(contracts/Demeter_CamelotFarm): Fix minor implementation issues.

    * Make Farm contract compatible with ERC721 receive.
    * Add PositionHelper interface.

commit 5cb56be167d25ae3dece0c99d50ce23d815f0bb3
Author: YashP16 <yash.pitroda786@gmail.com>
Date:   Wed Mar 29 14:17:48 2023 +0530

    feat(contracts/DemeterCamelotFarm): Add first draft for Demeter Camelot farms.

YashP16 committed:
refactor(contracts/Demter_E20_farm): Add base farm logic.

YashP16 committed:
fix(scripts): Merge error

YashP16 committed:
fix(contracts/Demeter_UniV3Farm): Resolve consistency issue in FARM_ID variable.

arcantheon committed:
Added balancer farm and deployer

arcantheon committed:
test script for create farm

arcantheon committed:
Added code doc

arcantheon committed:
Added code docs

arcantheon committed:
Added brownie utils back

arcantheon committed:
Added deployment script for balancer farm deployer

arcantheon committed:
Updated script to avoid failures

arcantheon committed:
refactor(Farm): Balancer farm to inherit base E20

arcantheon committed:
Added foundry

YashP16 committed:
Merge branch 'main' into wip/demeter-balancer

YashP16 committed:
Merge branch 'main' into dev

YashP16 committed:
Merge branch 'dev' into wip/demeter-balancer

arcantheon committed:
moved uniswap and balancer under e20 farms

arcantheon committed:
Added FARM_ID

bayou020 committed:
Initiated BaseFarm Test

bayou020 committed:
small update

parv3213 committed:
Improvements

parv3213 committed:
Improvements

YashP16 committed:
fix: Deployment artifacts file naming convention

YashP16 committed:
fix: Deployment artifacts file naming convention

parv3213 committed:
Make BaseFarm abstract

TechnoGeek01 committed:
used custom errors instead of require statements

parv3213 committed:
Add CI and husky. Other setup like USDsV2.

parv3213 committed:
Upper case constant variables

parv3213 committed:
Use forge fmt to format files
- Although skipping pre-commit for this commit, as formatted files may cause conflicts later

parv3213 committed:
Update solhint and fix solhint warnings

parv3213 committed:
Update CI

TechnoGeek01 committed:
optimized greater than zero checks for unsigned integers

TechnoGeek01 committed:
used unchecked in places where underflows are not possible. The if statements enclosing the calculations make sure they don't underflow

parv3213 committed:
Fix valid token check

parv3213 committed:
Remove `farmStartTime`

parv3213 committed:
Make `claimRewards(uint256 _depositId)` call `claimRewards(address _account, uint256 _depositId)` internally

parv3213 committed:
Add `_farmNotClosed` to `updateFarmStartTime` and `closeFarm`

TechnoGeek01 committed:
removed oldRewardRate from RewardRateUpdated event as it can be calculated from subgraph

TechnoGeek01 committed:
cached array length to save gas

TechnoGeek01 committed:
merged feat/improvements branch with feat/optimizations branch

parv3213 committed:
Lint fix

parv3213 committed:
Remove `getRewardTokens`

parv3213 committed:
Make `factory` immutable

parv3213 committed:
Merge branch 'feat/improvements' into feat/optimizations

TechnoGeek01 committed:
RewardAdded event in BaseFarm might require account which added the rewards. Added it as comments. Need to review

TechnoGeek01 committed:
Added depositId and removed tokenid,expirydate from CooldownInitiated event as they can be processed

TechnoGeek01 committed:
added depositId and removed tokenId,starttime in PoolUnsubscribed event as they can be processed

TechnoGeek01 committed:
optimized RewardsClaimed event to make it emit only once instead of twice in the _claimRewards function

TechnoGeek01 committed:
fixed rewardsForEachSubs nested array as mentioned in previous commit's comment. Also removed fundIds array from the RewardsClaimed event as fundId values are 0 and 1

TechnoGeek01 committed:
format of if statement condition is updated in updateFarmStartTime function for consistency. No change in functionality.

Co-authored-by: Parv Garg <parv3213@gmail.com>

TechnoGeek01 committed:
caching msg.sender will cost more gas than directly using it

TechnoGeek01 committed:
caching msg.sender will cost more gas than directly using it

arcantheon committed:
Added farm token check in recover erc20

bayou020 committed:
started testing

bayou020 committed:
Merge branch 'wip/demeter-balancer' into test/demeter-balancer
Merged from wip/demeter-balancer

bayou020 committed:
created farms

arcantheon committed:
Added two directories in gitignore

arcantheon committed:
Added contributing.md

arcantheon committed:
Added engineering dev room

arcantheon committed:
Updated README.md

arcantheon committed:
Added farm factory registration requirement

bayou020 committed:
added more test cases

bayou020 committed:
added coverage-minimum command

arcantheon committed:
Added farm id in e20 farm

arcantheon committed:
Updated discounted fees

arcantheon committed:
removed initialization of loop vars

arcantheon committed:
Update contracts/e20-farms/BaseE20Farm.sol

Co-authored-by: Parv Garg <parv3213@gmail.com>

arcantheon committed:
perf(deposit fn): code optimisation

arcantheon committed:
Merge branch 'wip/demeter-balancer' of github.com:Sperax/Demeter-Protocol into wip/demeter-balancer

arcantheon committed:
replaced memory var account with msg.sender

arcantheon committed:
Updated userDeposit from memory to storage

arcantheon committed:
Update variables storage spaces to save gas

bayou020 committed:
Added more test cases

bayou020 committed:
Added deposit tests

bayou020 committed:
Merge branch 'wip/demeter-balancer' into test/demeter-balancer

bayou020 committed:
(test:Balancer E20 Farms) Coverage 80%

TechnoGeek01 committed:
integrated sushiswap V3 into demeter

bayou020 committed:
added more BaseFarm tests

YashP16 committed:
Merge pull request #9 from Sperax/wip/demeter-balancer

Wip/demeter balancer

parv3213 committed:
Merge branch 'dev' into feat/improvements

parv3213 committed:
Fix failing build due to merge

YashP16 committed:
Merge pull request #11 from Sperax/feat/improvements

Feat/improvements

TechnoGeek01 committed:
change filename to avoid conflicts with windows

bayou020 committed:
Added Withdraw test cases

TechnoGeek01 committed:
removed a unnecessary comment in the RewardAdded emit

TechnoGeek01 committed:
merged dev branch into feat/optimizations branch

TechnoGeek01 committed:
used custom errors instead of require statements

TechnoGeek01 committed:
removed duplicate custom error

TechnoGeek01 committed:
reused files from uniswapV3

TechnoGeek01 committed:
Merge branch 'dev' into wip/demeter-sushiv3

TechnoGeek01 committed:
fixed arguments in Demeter_SushiV3FarmDeployer's constructor

TechnoGeek01 committed:
formatted Demeter_SushiV3FarmDeployer.sol to make it consistent with other files

TechnoGeek01 committed:
naming convention fix. CoolDown changed to Cooldown

TechnoGeek01 committed:
in RewardRateUpdated event marked rwdToken as indexed

TechnoGeek01 committed:
in CooldownInitiated event included indexed tokenid and expiry instead of depositid as tokenid is unique

YashP16 committed:
Merge pull request #12 from Sperax/feat/optimizations

Optimizations

arcantheon committed:
Merge branch 'dev' into chore/contributing-doc

bayou020 committed:
Merge branch 'dev' into test/demeter-balancer

arcantheon committed:
Added chore

TechnoGeek01 committed:
Merge branch 'dev' into wip/demeter-sushiv3

bayou020 committed:
Merge branch 'dev' into test/demeter-balancer

bayou020 committed:
Update .gitignore

Co-authored-by: Parv Garg <parv3213@gmail.com>

bayou020 committed:
Resolved @parv3213 Comments

bayou020 committed:
PR Fixes

parv3213 committed:
Fix package.json

parv3213 committed:
Format contracts

parv3213 committed:
Add husky to dev dependencies

bayou020 committed:
fix(Demeter Base Farm): Updated Test Cases

Updated Test Cases According to the new Contract Architecture,

BREAKING CHANGE:

bayou020 committed:
fix: removed Console from imports

bayou020 committed:
fix: Removed Console import from BaseFarm test file

parv3213 committed:
Fix file paths

parv3213 committed:
Format code

parv3213 committed:
Fix CI

parv3213 committed:
Fix gitignore

parv3213 committed:
forge install: forge-std

v1.7.1

parv3213 committed:
Simplify `foundry.toml`

parv3213 committed:
Fix file import

bayou020 committed:
test(Balancer Test Cases): Added Missing test cases

Added Claim Rewards and some view Functions test cases

TechnoGeek01 committed:
modularized sushi v3 and uni v3 codes

YashP16 committed:
fix: Resolve review comments

TechnoGeek01 committed:
sushiswapv3 FARM_ID name fix

Co-authored-by: Yash Pitroda <yash.pitroda786@gmail.com>

TechnoGeek01 committed:
uniswapv3 FARM_ID name fix

Co-authored-by: Yash Pitroda <yash.pitroda786@gmail.com>

TechnoGeek01 committed:
NPFM() return type fix

Co-authored-by: Yash Pitroda <yash.pitroda786@gmail.com>

TechnoGeek01 committed:
Revert "NPFM() return type fix"

This reverts commit c2cb9a9236b06c8d4964528c595622089ee4ae8f.

YashP16 committed:
Merge pull request #15 from Sperax/chore/contributing-doc

Open sourcing demeter contribution docs

YashP16 committed:
Merge pull request #16 from Sperax/wip/demeter-sushiv3

wip/demeter-sushiv3

parv3213 committed:
Merge remote-tracking branch 'origin/dev' into test/demeter-balancer

bayou020 committed:
fix(BaseFarm Test Files): Fixed Repeated GetPoolAddress Method on Test Files

parv3213 committed:
Fix forge lib and removing unnecessary file

parv3213 committed:
Fix solhint errors

parv3213 committed:
forge install: forge-std

v1.7.1

parv3213 committed:
Merge remote-tracking branch 'origin/dev' into fix/formatting

parv3213 committed:
Fix lint error

parv3213 committed:
Add solhint checks to CI

parv3213 committed:
Remove prettier dependencies

parv3213 committed:
Update PR-CI and and

parv3213 committed:
Format code from `prettier` script

YashP16 committed:
Merge pull request #13 from Sperax/fix/formatting

Use forge fmt to format files

parv3213 committed:
Merge remote-tracking branch 'origin/dev' into test/demeter-balancer

parv3213 committed:
Format code using `prettier` script

parv3213 committed:
Turn `contract-name-camelcase` off in solhint

parv3213 committed:
Fix solhint warnings

parv3213 committed:
Remove comment

YashP16 committed:
refactor(contracts): Make initiateCooldown and withdraw() accessible from BaseFarm

YashP16 committed:
test(utils): Add network configuration for tests

YashP16 committed:
test: Restructure test cases to make them modular

YashP16 committed:
test: Remove old test cases

parv3213 committed:
Fix import path in BaseE20Farm.t.sol

YashP16 committed:
test: Add setup modifier for tests

bayou020 committed:
feat(BaseFarm): Added a Non Zero Amount Check

Added a non Zero amount check in Add Rewards function

bayou020 committed:
test(BalancerFarm): Fixed not working test cases

bayou020 committed:
Solhint Functions ordering fixes

bayou020 committed:
fix(Commit Conflicts): Fixed Commit  Conflicts on test Cases

parv3213 committed:
Multiple fixes

- Uniswap pool initialization order in
BaseUniV3Farm
- Make `NFPM` and `UNIV3_FACTORY` public view
- Add required interfaces for UniswapV3
- Rename `Demeter_UniV3FarmDeployer` to `Demeter_SushiV3FarmDeployer`

parv3213 committed:
Add UniV3 Farm tests

bayou020 committed:
feat(Fuzzing tests): added fuzzing amounts tests

parv3213 committed:
Rename by adding `.t.sol` to tests

bayou020 committed:
feat(Balancer Farm): Added Test Cases Imports

bayou020 committed:
feat(Base Farm): Added Missing test cases

bayou020 committed:
feat(BaseE20Farm): Added Reverting Test Cases

arcantheon committed:
Updated BaseFarmDeployer

arcantheon committed:
Updated other deployers

arcantheon committed:
Updated farm factory

arcantheon committed:
Updated param in function getFees

bayou020 committed:
feat(BaseSetup): Fixed Owner= User Bug, Added More Actors , Added Factory Variable

bayou020 committed:
fix(BaseFarm): Owner actor shouldn't be user

bayou020 committed:
fix(FarmFactory): Added test cases for Farm Factory

bayou020 committed:
fix(BaseE20Farm): Fixed test cases when User Prank is not owner

bayou020 committed:
fix(BaseFarm): Minor Prank Fix

parv3213 committed:
Check for `_isNonZeroAmt` before `_farmNotClosed`

parv3213 committed:
Fix/improve tests

bayou020 committed:
fix(FarmFactory test cases): Solved Mehul's comments

arcantheon committed:
Updated Readme

parv3213 committed:
Add Multicall functionality to BaseFarm contract with tests

bayou020 committed:
fix(BaseFarm.t.sol ): Fix Parv's comments

- Fixed RevertsWhen Naming convetion
- Fixed ExpectEmit Failed assertions
- Added Maths tests

bayou020 committed:
fix(BaseE20Farm.t.sol ): Added Maths tests

Added Maths tests for increase Deposit and withdraw Partially functions

bayou020 committed:
fix(BalancerFarn.t.sol ): Fixed Parv's Comment

- Added a check on Farm ID
- Added Deposit Maths check

bayou020 committed:
fix(BaseFarm.t.sol ): Typo fix

test_deposit_noLockupFarm_revertsWhen_NoLiquidityInPosition fix

bayou020 committed:
fix(BaseFarm.t.sol BaseFarm.t.sol): Fixes Parv's Comments

bayou020 committed:
fix(BaseFarm.sol): Removed 0 amount Function check

YashP16 committed:
Merge pull request #17 from Sperax/test/demeter-balancer

Test/demeter balancer

parv3213 committed:
Merge remote-tracking branch 'origin/dev' into fix/solhint-warnings

parv3213 committed:
Fix solhint errors in tests

parv3213 committed:
Merge remote-tracking branch 'origin/dev' into test/demeter-uniV3Farm

parv3213 committed:
Fix tests failing after merge

YashP16 committed:
Merge pull request #18 from Sperax/fix/solhint-warnings

Fix solhint warnings

arcantheon committed:
getFeeParams external call updates

mehultuteja committed:
Merge branch 'dev' into wip/deployer-upgrades

mehultuteja committed:
Fixed test cases

YashP16 committed:
Merge pull request #22 from Sperax/wip/deployer-upgrades

Remove discounted fee calculation logic from Deployers and Farm factory updates

bayou020 committed:
feat(BaseFarm.t.sol): Fixed ERC20Recover Coverage, Added two edge cases

- Fixed erc20Recover coverage
- Added two claim rewards edge cases

bayou020 committed:
feat(BaseE20Farm.t.sol): Added ERC20Recover Specific Logic

bayou020 committed:
feat(BalancerFarm.t.sol): Fixed ERC20Recover Specific Import

bayou020 committed:
feat(SushiV3Farm.t.sol): added ERC20Recover Import

parv3213 committed:
Merge remote-tracking branch 'origin/dev' into test/demeter-uniV3Farm

parv3213 committed:
Resolve solhint errors

parv3213 committed:
Merge remote-tracking branch 'origin/dev' into test/demeter-uniV3Farm

parv3213 committed:
Rearrange _SetupFarmTest in BalancerFarm.t.sol

parv3213 committed:
Fix stack-too-deep error by adding NonfungiblePositionManagerUtils

- https://github.com/Sperax/Utility/pull/2

parv3213 committed:
Merge remote-tracking branch 'origin/dev' into feat/multicall

parv3213 committed:
Add more MultiCall tests

parv3213 committed:
Add tests supporting BaseFarm coverage

bayou020 committed:
feat(BaseFarm): Remove Redundant check

- Remove Redundant check on  _subscribeRewardFund  & _unsubscribeRewardFund

YashP16 committed:
Merge pull request #23 from Sperax/feat/multicall

Add Multicall to BaseFarm

bayou020 committed:
Merge branch 'dev' into test/demeter-uniV3Farm
- Fixing the conflicts
-Fixing Linting Conflicts

parv3213 committed:
Refactor: Remove `uniswapUtils` & `nfpmUtils` from `UniswapPoolData`.

parv3213 committed:
Remove coverage-minimum script from package.json

bayou020 committed:
test(BaseFarm): Added Assertions

Added Claim Rewards Balances Assertions

YashP16 committed:
Merge pull request #20 from Sperax/test/demeter-uniV3Farm

Demeter UniswapV3 (UniswapV3 and SushiSwap) Farm Tests

TechnoGeek01 committed:
updated vault address in Arbitrum.t.sol (#30)

arcantheon committed:
Unique deposit ID and event data optimisation (#27)

* Code changes for unique depositId

* Updated test cases

* Added indexed parameter depositId

* Added a view function to get deposits of a user

* Removed getMyDeposit and deposit updates

* event data optimizations

* Code doc improvements

* Test cases compilation error fixes

* subscription to depend on deposit id

* Added deposits assignment before sub reward fund

* fix(deposit): storage write instead of memory

* Updated test cases

* fixed events in test files

* fix(event data): totalRewardsClaimed to refer from storage instead of memory

* Removed logs

* Added revert in getDeposit function

* updated test cases as per latest change in code

* fixed test cases

* Added PoolSubscribed event

* added miscellaneous test cases to check if one user's withdrawal of deposit affects other user's deposit

* added test cases in initiatecooldowntest to make sure someone else other than the depositor cannot initiate cooldown

* added missing emit event tests

* added missing event test in test_initiateCooldown_LockupFarm test case.

* added test cases to make sure one user's withdraw doesn't affect other users' subscriptions and deposits

* fixed a test case

* Gas optimisations

* Fixed review comments

* updated variable name

* Removed tokenId and simplified further

* Fixed test cases

* renamed mapping and updated _deposit function

* updated variable names instead of integer literals

* renamed test function names

* fixed a redundancy in test case

* optimized event testing

* fixed a event testing

* Updated updateTokenManager function and event name

* gas optimisations and code improvements

* Gas optimisation and function name change

* renamed updateRewardData function name

* Added mistakenly removed test case back

---------

Co-authored-by: mehultuteja <reachmehultuteja@gmail.com>
Co-authored-by: TechnoGeek01 <rsudharsangames@gmail.com>
Co-authored-by: YashP16 <yash.pitroda786@gmail.com>

parv3213 committed:
ci(github): Update CI to report gas difference (#34)

Using Rubilmax/foundry-gas-diff

parv3213 committed:
Revert "ci(github): Update CI to report gas difference (#34)" (#35)

This reverts commit 778bdc853ac4913ae48a64b8ceabad21e468af80.

arcantheon committed:
Feature to increase/ decrease liquidity from Camelot position while being staked in Demeter. (#26)

* Added increase and decrease liquidity func

* Fixed deployment and base errors

* Fixed errors

* Fixed all the pending test cases

* Added onNFTAddToPosition

* Added a view function for getting amounts

* Added onNFTWithdraw function

* test(BaseFarm): Added DepositIncreased/Decreased Events

* fix(CamelotFarm): Small Bug Fix

* test(CamelotFarm): Added test cases

Increase/ Decrease Deposit test cases

* Updated test case to fix failure

* test(CamelotFarm): Added test cases

Increase/ Decrease Deposit test cases

* Merge branch 'dev' into feat/camelot-farm-liq

* test(CamelotFarm): Added test cases

Increase/ Decrease Deposit test cases

* test(CamelotFarm): Added Test Cases

- Coverage is 98%
- Checked Maths on increase / Decrease Deposit

* Added Camelot Addresses

* fix(CamelotFarm.t.sol): Fixed Assets Naming

- Fixed Assets Naming
- Removed Constants from the test file

* Fix RecoverERC20E20FarmTest

* Updated Camelot interface file name

* Update contracts/camelot/Demeter_CamelotFarm.sol

Co-authored-by: Parv Garg <parv3213@gmail.com>

* Added update call and removed comments

* Removed redundant calls

* fix(Fix commented Tests): Fixed commented tests

* Updated test cases

* Refactor removeLiquidity function to use named parameters

* Moved subscriptions updater in abstract contract

* Updated tests for OperableDeposit changes

* updated docs for OperableDeposit changes

* fix(Test Cases): Fix function names to use proper capitalization

* perf(test): approval and naming conventions

* perf(Farms): approval function call and naming convention

* Fix function names to use proper capitalization

* Refactor deposit tests in Demeter_CamelotFarmTest

* refactor(CamelotFarm Testing): refractored test cases

* Add GetDepositTest and MulticallTest to DemeterCamelotFarmInheritTest

* Remove unused test cases in BaseE20Farm.t.sol

---------

Co-authored-by: mehultuteja <reachmehultuteja@gmail.com>
Co-authored-by: Dayaa <m.d.e.boutalbi@me.com>
Co-authored-by: Parv3213 <parv3213@gmail.com>

parv3213 committed:
ci(github): Update CI to report gas difference (#36)

Using Rubilmax/foundry-gas-diff

TechnoGeek01 committed:
feat/basefarm-upgrades (#25)

* added farm expiry and extension feature with extension fees

* standardization

* reordered functions as per solhint standards and errors

* fixed a bug in extendFarmEndTime function

* fixed a bug in extendFarmEndTime function

* updated function name for collecting extension fee

* optimization

* added some test cases

* updated test case

* added farm extension test cases

* added ExtensionFeeCollected event test case

* optimized updateFarmStartTime function

* optimized FarmEndTimeUpdated emit

* changed visibility of _collectExtensionFee in BaseFarm as private

* added constant for initial farm extension

* removed getFees function in BaseFarmDeployer as it's redundant

* removed FarmEndTimeUpdated emit in updateFarmStartTime function as it can be processed until farm starts

* added nonReentrant modifier to extendFarmEndTime function

* optimization

* fixed test cases

* function reordering to fix solhint errors

* added some missing assertions in FarmFactory.t.sol test cases

* removed unnecessary logs and comments

* event optimization

* updateFarmStartTime function optimization

* destructured farmInstance.iniitialize parameters

* extendFarmEndTime function gas optimization

* renamed the function extendFarmEndTime to extendFarmDuration

* removed creator field in ExtensionFeeCollected event

* fixed a bug

* renamed the function _farmNotClosedOrExpired to _farmNotClosedOrNotExpired

* added more test cases

* added some missing closeFarm test cases.

* updated missing natspec

* renamed the function _farmNotClosedOrNotExpired to _isFarmActive

* removed unnecessary fuzzing

* updated usds vault address to the latest one

* added internal function for farm expiry validation

* added test cases for some edge cases

* decoupled farm expiry feature in a separate abstract contract

* fixed solhint warnings

* decoupled farm expiry status function

* comment cleanup

* decoupled farmFactory state variable

* fixed some solhint errors

* fixed solhint warnings

* moved FarmNotYetStarted from BaseFarm to BaseFarmExpiry

* improved test cases

* made fee variables configurable in FarmFactory.t.sol

* fixed test_claimRewards_max_rewards test case

* renamed contract BaseFarmExpiry to BaseFarmWithExpiry

* removed logs

* added internal function _updateFarmStartTime

* update

* added missing test inheritances

* natspec update

* abstracted farm expiry test cases

* added constants instead of hardcoded values

* update

* Natspec update

Co-authored-by: arcantheon <140178288+arcantheon@users.noreply.github.com>

* Natspec update

Co-authored-by: arcantheon <140178288+arcantheon@users.noreply.github.com>

* added zero checks in updateFeeParams function of FarmFactory.sol

* updated natspec

* Revert "added zero checks in updateFeeParams function of FarmFactory.sol"

This reverts commit 9c080d346e00f3073efb9f726504ca49caae6107.

* added some missing test cases

* fixed fuzz runs in foundry.toml

* expiry test cases restructure

* updated natspec

* updated natspec

* updated natspec

* test case optimization and redundancy removal

* improved test cases

---------

Co-authored-by: arcantheon <140178288+arcantheon@users.noreply.github.com>

parv3213 committed:
Draft changes for chain agnostic contracts (#32)

* Draft changes for chain agnostic contracts

* fix(contracts): Fix problems

* fix: Bug fixes

* fix: Optimisation fixes

* fix(tests): fix failing tests

* fixed BaseE20Farm.t.sol test cases

* test fix

* fixed Demeter_CamelotFarm.t.sol test cases

---------

Co-authored-by: TechnoGeek01 <rsudharsangames@gmail.com>

arcantheon committed:
Storage, DataTypes separation and OperableDeposit simplification (#38)

* Storage, DataTypes separation and OperableDeposit simplification

* updated test cases to support separated DataTypes

* Usage of storage variables for gas optimization

* Using named imports instead of default

* abstracted contract BaseFarmStorage

* Moved OperableDeposit under features

* Removed constant parameters

TechnoGeek01 committed:
Restructured computeRewards function's return data in BaseFarm.sol (#39)

* restructured computeRewards return data

* updated natspec

TechnoGeek01 committed:
function naming convention fix and restructuring (#40)

* function naming convention changes

* restructured functions

* updated natspec

* bug fix

* functions rename

* function restructuring

* unused variable cleanup in test file

* function restructuring

parv3213 committed:
Demeter UniswapV3 Farm upgrades (#21)

* Add liquidity management functions to UniswapV3
interface

* Remove DepositInCooldown error and reuse
DepositIsInCooldown

* Add increaseDeposit and withdrawPartially
functions for UniV3 Farm

* Add ActiveLiquidityBaseV3Farm

- Farm that distributes rewards only for active liquidity
- Withdraw from locked position is allowed when liquidity is not active

* Fix tests after merge

* Rename `ActiveLiquidityBaseUniV3Farm` to `BaseUniV3ActiveLiquidityFarm`

* refactor(BaseUniV3Farm): Use NFPM utils to resolve stack-too-deep error

* test(BaseUniV3): add tests for increase/decrease liquidity

* test: bugfix and tests for BaseUniV3ActiveLiquidityFarm

* refactor: modularize BaseFarm and reduce duplicate code in BaseUniV3ActiveLiquidityFarm

* refactor and natspec fix

* Optimise contracts/uniswapV3/BaseUniV3ActiveLiquidityFarm.sol

Co-authored-by: Sudharsan R <34187497+TechnoGeek01@users.noreply.github.com>

* test: add missing test

* bug fix

* refactor: remove unchecked

* Update lastSecondsInside data type to uint32

* Update contracts/camelot/Demeter_CamelotFarm.sol

Co-authored-by: arcantheon <140178288+arcantheon@users.noreply.github.com>

* Update contracts/camelot/Demeter_CamelotFarm.sol

Co-authored-by: arcantheon <140178288+arcantheon@users.noreply.github.com>

* Refactor deposit token ID usage in Demeter_CamelotFarm.sol

---------

Co-authored-by: Sudharsan R <34187497+TechnoGeek01@users.noreply.github.com>
Co-authored-by: arcantheon <140178288+arcantheon@users.noreply.github.com>

arcantheon committed:
Abstracting NFT farms (BaseE721Farm.sol) (#42)

* refactor(farms): moved initiateCooldown to BaseFarm

* refactor(camelot): uncommented interface and override in contract

* Removed unused variables

* Added BaseE721Farm.sol

* Updated tests to reflect nftContract name update

* Moved common functionality to base and documentation update

* Updated test cases to support the changes

* moved _validateDeposit check under _withdraw()

* updated inheritance hierarchy

* updated natspec and removed virtual keyword

* Moved camelot and uniV3 under e721 farms

* moved nft tests under e721-farms and added BaseE721Farm

* removed unneeded typecasting

* refactored tests and moved common ones under BaseE721Farm.t.sol

* fixed imports related issues

* Merge branch 'dev' into feat/e721-farms

* fix: Fix imports

* fix: Fix imports

---------

Co-authored-by: mehultuteja <reachmehultuteja@gmail.com>
Co-authored-by: Parv3213 <parv3213@gmail.com>

TechnoGeek01 reviewed (approved):
* "FarmFactory" contract can be renamed to "FarmRegistry"

parv3213 committed:
BaseFarm Optimizations (#33)

* perf(BaseFarm): Optimizations

* perf(BaseFarm): Optimizations

TechnoGeek01 committed:
Test case optimization (#41)

* test case optimization

* test case optimization

* test case optimization

* fix

* naming convention fix

* removed unnecessary fuzz in negative test cases

* rename and restructure of test cases

* removed asserts and emit tests from deposit helper function and wrote separate tests

* helper function optimization

* added camelot deposit test

* cleanup

* added missing test contract inheritance

* rename

* remove unnecessary fuzz for negative cases

* fuzz functions rename

* update

* test optimization

* optimization

* optimization

* cleanup

* optimization

* rename

* optimization

* rename

* cleanup

* update

* rename revert tests for consistency

* refactor(test): Remove redundant code from uniV3 test files

* refactor(contracts/BaseE20Farm): Remove redundancy in `recoverERC20` function

* refactor(test): Simplify test hierarchy

---------

Co-authored-by: Parv3213 <parv3213@gmail.com>
Co-authored-by: Yash Pitroda <yash.pitroda786@gmail.com>

arcantheon committed:
fix(BaseFarm._setupFarm): Updated check for rwdTokens.length (#47)

* fix(BaseFarm._setupFarm): Fixed the check for rwdTokens.length

* Updated test cases for reward token data check bug fix

---------

Co-authored-by: mehultuteja <reachmehultuteja@gmail.com>

TechnoGeek01 committed:
optimized cooldown period (#45)

* optimized cooldown period

* refactor

Co-authored-by: Parv Garg <parv3213@gmail.com>

* updated natspec

* fix

* rename

* restructure

* updated _validateCooldownPeriod

* variable name simplification

* optimization

---------

Co-authored-by: Parv Garg <parv3213@gmail.com>

TechnoGeek01 committed:
code reuse optimization. (#48)

* optimization

* optimization

* rename

* optimization

* removed redundant check and did optimization

* rename

* decrease deposit further abstraction

* increase deposit further abstraction

* fix

TechnoGeek01 committed:
rename of contracts, files, variables, etc (#46)

* renamed BaseFarmWithExpiry to ExpirableFarm

* Removed Base from all the farm names

* renamed factory to registry

* rename

* update

* renamed camelot farms similar to univ2 and univ3

* natspec update

* natspec update

* removed forge docs

* removed Demeter_ from farm names

* added Demeter ascii logo in our contracts

* name fix

* rename

* name standardization

arcantheon committed:
Enabling fixed APR reward tokens for Demeter Protocol (#50)

* Initial version of DemeterRewarder contract

* Added fixed APR struct in DataTypes

* Added farmRegistered in farm registry's interface

* Added USDs oracle's interface

* Added apr description

* Demeter rewarder contract all functionalities

* First draft of DemeterRewarder

* Updated IFarmFactory -> IFarmRegistry

* Updated logo

* Added getTokenAmounts function for DemeterRewarder

* Implemented getTokenAmounts for CamelotV2

* Refactored from reward pool to individual rewarder

* Added natspec

* Implemented getTokenAmounts for univ2 and univ3

* Added call to _transferOwnership()

* Added rewards budget functionality

* Added a sample test script to test the functionality

* removed redundant variable declaration

* Removed farm registered function from interface

* removed redundant transfer ownership call

* Reconfigured _normalizeAmount to cache decimals

* Updated isFarmValid check

* Removed adjustment of rewardsPerSec

* Updated variable names introduced a constant for 1 weeks

* Added to farm validation logic

* Updated naming convention

* Made E20Farm abstract

* Fixed rewardRate bug

* Renamed struct and event name

* Added FixedAPRRewarderSPA

* Final version of FixedAPRRewarderSPA

* Rewarder improvisation updated and bug fixes

* Fixed APR rewarder for Legacy farms with simulation script

* Added recover erc20 and updated calibrateRewards

* Added total reward rate

* Added FarmRewardConfigParams

* Fixed failing test case

* Cached baseTokens indexes in farm's asset tokens

* Updated test script

* file restructuring, improved naming

* Arch change in Rewarder and legacy rewarder

* Added getRewardTokens test

* Added _isConfigured check and updateAPR function

* Updated function name

* Naming convention improvement and gas optimization.

* Update contracts/rewarder/RewarderFactory.sol

Co-authored-by: Parv Garg <parv3213@gmail.com>

* Update contracts/rewarder/RewarderFactory.sol

Co-authored-by: Parv Garg <parv3213@gmail.com>

* Added nonReentrant and made rewardTokens internal

* Added Average price consultation instead of spot

* Updated Uniswap utils

* Added test cases for rewarder

* Added isConfigured check in calibrateReward

* feat(calibrateReward): Added calibration restriction functionality

* Added a dependency comment

* Added update config test cases

---------

Co-authored-by: mehultuteja <reachmehultuteja@gmail.com>
Co-authored-by: Parv Garg <parv3213@gmail.com>

TechnoGeek01 committed:
optimized Uni V3 claim fee (#54)

* optimized claim fee

* asserted claim swap fee for uniswap v3

* added asserts for claim swap fee

* named imports

TechnoGeek01 committed:
Camelot V3 (#52)

* added Camelot V3

* initial test setup

* added camelot v3 nfpm utils contract address

* test fix

* added test cases

* update

* optimization -> removed unnecessary dependency contract

* update

* tested edge cases -> farm functionality after tickspacing change

* update

* fix

* update

* cleanup

* removed unnecessary nfpmutils contract

* Revert "removed unnecessary nfpmutils contract"

This reverts commit 1a53eb88ba3f33278e8216bd10e437030183c01e.

* reintroduced camelot utils as it's required later

* update

* added rewarder functionality for camelot v3

* fix

* updated camelot utils contract address

* cleanup

* removed unused package

* rename

* cleanup

* cleanup

* added gettokenamounts test

* update

* improved camelot fee test case

* removed redundant test cases

* code reuse

* removed redundant code

* updated initialize with a single struct param

* named imports

* natspec update

* natspec update

---------

Co-authored-by: Yash Pitroda <yash.pitroda786@gmail.com>

arcantheon committed:
Perf/solidity version to 0.8.24 (#55)

* Updated openzeppelin version

* Upgraded version and fixed ReentrancyGuard path

* Ownable constructor param added

* Updated UniV3 initialize params

* Updated solhint compiler version

* Fixed breaking changes and solhint warnings

* Removed underscores from struct member's definition

* perf: minor changes

* Updated OZ dependencies

---------

Co-authored-by: Parv3213 <parv3213@gmail.com>

TechnoGeek01 committed:
fixed a test case (#56)

* fixed a test case

* update naming

TechnoGeek01 committed:
testing improvement (#59)

* fixed a camelot v3 test case

* test optimization

arcantheon committed:
Test/demeter rewarder (#57)

* Initial version of DemeterRewarder contract

* Added fixed APR struct in DataTypes

* Added farmRegistered in farm registry's interface

* Added USDs oracle's interface

* Added apr description

* Demeter rewarder contract all functionalities

* First draft of DemeterRewarder

* Updated IFarmFactory -> IFarmRegistry

* Updated logo

* Added getTokenAmounts function for DemeterRewarder

* Implemented getTokenAmounts for CamelotV2

* Refactored from reward pool to individual rewarder

* Added natspec

* Implemented getTokenAmounts for univ2 and univ3

* Added call to _transferOwnership()

* Added rewards budget functionality

* Added a sample test script to test the functionality

* removed redundant variable declaration

* Removed farm registered function from interface

* removed redundant transfer ownership call

* Reconfigured _normalizeAmount to cache decimals

* Updated isFarmValid check

* Removed adjustment of rewardsPerSec

* Updated variable names introduced a constant for 1 weeks

* Added to farm validation logic

* Updated naming convention

* Made E20Farm abstract

* Fixed rewardRate bug

* Renamed struct and event name

* Added FixedAPRRewarderSPA

* Final version of FixedAPRRewarderSPA

* Rewarder improvisation updated and bug fixes

* Fixed APR rewarder for Legacy farms with simulation script

* Added recover erc20 and updated calibrateRewards

* Added total reward rate

* Added FarmRewardConfigParams

* Fixed failing test case

* Cached baseTokens indexes in farm's asset tokens

* Updated test script

* file restructuring, improved naming

* Arch change in Rewarder and legacy rewarder

* Added getRewardTokens test

* Added _isConfigured check and updateAPR function

* Updated function name

* Naming convention improvement and gas optimization.

* Update contracts/rewarder/RewarderFactory.sol

Co-authored-by: Parv Garg <parv3213@gmail.com>

* Update contracts/rewarder/RewarderFactory.sol

Co-authored-by: Parv Garg <parv3213@gmail.com>

* Added nonReentrant and made rewardTokens internal

* Added Average price consultation instead of spot

* Updated Uniswap utils

* Added test cases for rewarder

* Added isConfigured check in calibrateReward

* feat(calibrateReward): Added calibration restriction functionality

* Added a dependency comment

* Added update config test cases

* Merge branch 'dev' into test/Demeter-rewarder

* Completed rewarder test cases

* variable naming updated

* Updated var names

---------

Co-authored-by: mehultuteja <reachmehultuteja@gmail.com>
Co-authored-by: Parv Garg <parv3213@gmail.com>
Co-authored-by: Yash Pitroda <yash.pitroda786@gmail.com>

TechnoGeek01 committed:
Natspec update for demeter protocol. (#58)

* updated natspec

* updated natspec for TokenUtils.sol

* correction

* update

* update

* natspec consistency update

* Update contracts/e20-farms/E20Farm.sol

Co-authored-by: Parv Garg <parv3213@gmail.com>

* Update contracts/e721-farms/E721Farm.sol

Co-authored-by: Parv Garg <parv3213@gmail.com>

* Update contracts/e721-farms/camelotV2/CamelotV2Farm.sol

Co-authored-by: Parv Garg <parv3213@gmail.com>

* added demeter logo in TokenUtils.sol

* update

* update

---------

Co-authored-by: Parv Garg <parv3213@gmail.com>

arcantheon committed:
Farm optimisations (#60)

* Optimised Farm

* Updated test cases

* Update contracts/Farm.sol

Co-authored-by: Parv Garg <parv3213@gmail.com>

* Update contracts/Farm.sol

---------

Co-authored-by: Parv Garg <parv3213@gmail.com>

bayou020 commented on contracts/e20-farms/uniswapV2/UniV2FarmDeployer.sol:
```diff
@@ -0,0 +1,110 @@
+    // farmStartTime - Timestamp when reward accrual begins for deposits in the farm.
+    // cooldownPeriod - Cooldown period for locked deposits (in days).
+    //                  Make cooldownPeriod = 0 for disabling lockup functionality of the farm.
+    // camelotPoolData - data for camelot pool (tokenA, tokenB).
```
Typo here It's uniFarmData @YashP16

parv3213 committed:
feat: Use OpenZeppelin SafeCast for preventing unexpected cases (#62)

parv3213 committed:
Audit Fixes: QS-1 (#61)

* feat: Add failing test for deposit before farm start time

* bug: if `lastFundUpdateTime` is greater than block time, return 0. Instead of underflow.

* fix: Fix multiple time related problem:

1. Add `farmStartTime` in place of `lastFundUpdateTime`
2. Change logic when `lastFundUpdateTime` is updated. This is added to fix problems in UniV3ActiveFarm
3. Fixed testcases for the changes

* chore: better comments

* chore: rename `_currentFarmStartTime` to `currentFarmStartTime`

* bug: add check if `lastFundUpdateTime == 0`, if so return 0.

This will update lastFundUpdateTime = block.timestamp in the same transaction, and will start accruing rewards from from the next tx onwards

parv3213 committed:
Audit Fixes: QS-9 (#63)

* refactor: Update privilege address mapping name in FarmRegistry.sol for clarity

* chore: rename `_userAddress` to `_user`

---------

Co-authored-by: Yash Pitroda <yash.pitroda786@gmail.com>

arcantheon committed:
fix(Rewarder): consider acculumated rewards instead of balance (#65)

Co-authored-by: Yash Pitroda <yash.pitroda786@gmail.com>
